### PR TITLE
IGNITE-10544: GMM with fixed components

### PR DESCRIPTION
**Pull request by avplatonov**

### CODE REVIEW

File: modules/ml/src/main/java/org/apache/ignite/ml/math/primitives/matrix/AbstractMatrix.java [hunk 1]
```diff
@@ -918,4 +918,38 @@ else if (mean > threshold + iv)
     @Override public String toString() {
         return "Matrix [rows=" + rowSize() + ", cols=" + columnSize() + "]";
     }
+
+    /** {@inheritDoc} */
+    @Override public double determinant() {
+        //TODO: IGNITE-11192, use nd4j
+        LUDecomposition dec = new LUDecomposition(this);
+        double res = dec.determinant();
+        dec.destroy();
```
Comment (dmitrievanthony) [inline modules/ml/src/main/java/org/apache/ignite/ml/math/primitives/matrix/AbstractMatrix.java]: Does it make sense to wrap it into `try-finally`?

File: modules/ml/src/main/java/org/apache/ignite/ml/math/stat/DistributionMixture.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+/**
+ * Mixture of distributions class where each component has own probability and probability of input vector can be
+ * computed as a sum of likelihoods of each component.
+ *
+ * @param <C> distributions mixture component class.
+ */
+public abstract class DistributionMixture<C extends Distribution> implements Distribution {
+    /** */
+    private final double EPS = 1e-5;
```
Comment (dmitrievanthony) [inline modules/ml/src/main/java/org/apache/ignite/ml/math/stat/DistributionMixture.java]: Would be great to add `static`.

File: modules/ml/src/main/java/org/apache/ignite/ml/nn/ReplicatedVectorMatrix.java [hunk 3]
```diff
@@ -551,4 +551,18 @@ public boolean isColumnReplicated() {
...
         return vector;
     }
+
+    /** {@inheritDoc} */
+    @Override public double determinant() {
+        // If matrix is not square throw exception.
+        checkCardinality(vector.size(), replicationCnt);
+
+        // If matrix is 1x1 then determinant is its single element otherwise there are linear dependence and determinant is 0.
+        return vector.size() > 0 ? 0 : vector.get(1);
```
Comment (dmitrievanthony) [inline modules/ml/src/main/java/org/apache/ignite/ml/nn/ReplicatedVectorMatrix.java]: Looks weird, isn't it?

**Comment by dmitrievanthony** [comment]
Please, add a test for the case when cache partition is empty.

File: modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/CovarianceMatricesAggregator.java [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    private static final long serialVersionUID = 4163253784526780812L;
+
+    /** Mean vector. */
+    private final Vector mean;
+
+    /** Weighted by P(c|xi) sum of (xi - mean) * (xi - mean)^T values. */
+    private Matrix weightedSum;
+
+    /** Count of rows. */
+    private int N;
```
[the remark this answers is not in the document]
Comment (avplatonov) [reply modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/CovarianceMatricesAggregator.java]: done

File: modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/CovarianceMatricesAggregator.java [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+    }
+
+    /**
+     * Computes covatiation matrices for feature vector for each GMM component.
+     *
+     * @param dataset Dataset.
+     * @param clusterProbs Probabilities of each GMM component.
+     * @param means Means for each GMM component.
+     */
+    public static List<Matrix> computeCovariances(Dataset<EmptyContext, GmmPartitionData> dataset,
```
[the remark this answers is not in the document]
Comment (avplatonov) [reply modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/CovarianceMatricesAggregator.java]: I want to group such methods to one logic unit like statistics computer and avoid to create a lot of little objects with specific functionality that can't be interpreted without other class.

[the remark this answers is not in the document]
Comment (avplatonov) [reply modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/CovarianceMatricesAggregator.java]: I change visibility to package, but I won't move this methods to other classes

File: modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/CovarianceMatricesAggregator.java [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+     * @return computed covariance matrix.
+     */
+    private Matrix covariance(double clusterProb) {
+        return weightedSum.divide(N * clusterProb);
+    }
+
+    /**
+     * Reduce stage for covariance computation over dataset.
+     *
+     * @param l L.
```
[the remark this answers is not in the document]
Comment (avplatonov) [reply modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/CovarianceMatricesAggregator.java]: :-)

File: modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/GmmModel.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+        for (int i = 0; i < likelihood.size(); i++) {
+            double prob = likelihood.get(i);
+            if (prob > maxLikelihood) {
+                maxLikelihood = prob;
+                idxOfMax = i;
+            }
+        }
+
+        return (double)idxOfMax;
```
[the remark this answers is not in the document]
Comment (avplatonov) [reply modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/GmmModel.java]: :+1:

File: modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/GmmPartitionData.java [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+        public Builder(FeatureLabelExtractor<K, V, Double> extractor, int countOfComponents) {
+            this.extractor = extractor;
+            this.countOfComponents = countOfComponents;
+        }
+
+        /** {@inheritDoc} */
+        @Override public GmmPartitionData build(LearningEnvironment env, Iterator<UpstreamEntry<K, V>> upstreamData,
+            long upstreamDataSize, EmptyContext ctx) {
+
+            List<LabeledVector<Double>> xs = new ArrayList<>((int)upstreamDataSize);
```
[the remark this answers is not in the document]
Comment (avplatonov) [reply modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/GmmPartitionData.java]: :+1:

File: modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/GmmPartitionData.java [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+                pcxi[counter] = new double[countOfComponents];
+                counter++;
+            }
+
+            return new GmmPartitionData(xs, pcxi);
+        }
+    }
+
+    /**
+     * Sets P(c|xi) = 1 for closest cluster "c" for each vector in partition data using initial means as cluster centers
```
[the remark this answers is not in the document]
Comment (avplatonov) [reply modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/GmmPartitionData.java]: remove older comment with "return"

File: modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/GmmPartitionData.java [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+            long upstreamDataSize, EmptyContext ctx) {
+
+            List<LabeledVector<Double>> xs = new ArrayList<>((int)upstreamDataSize);
+            double[][] pcxi = new double[(int)upstreamDataSize][];
+            int counter = 0;
+            while (upstreamData.hasNext()) {
+                UpstreamEntry<K, V> entry = upstreamData.next();
+                LabeledVector<Double> x = extractor.extract(entry.getKey(), entry.getValue());
+                xs.add(x);
+                pcxi[counter] = new double[countOfComponents];
```
[the remark this answers is not in the document]
Comment (avplatonov) [reply modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/GmmPartitionData.java]: :+1:

File: modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/GmmPartitionData.java [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+     */
+    public int size() {
+        return pcxi.length;
+    }
+
+    /**
+     * @return count of GMM components.
+     */
+    public int countOfComponents() {
+        return pcxi[0].length;
```
[the remark this answers is not in the document]
Comment (avplatonov) [reply modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/GmmPartitionData.java]: :+1:

File: modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/GmmTrainer.java [hunk 12]
```diff
@@ -0,0 +1,346 @@
...
+    }
+
+    /**
+     * Sets numberOfComponents.
+     *
+     * @param numberOfComponents Number of components.
+     * @return trainer.
+     */
+    public GmmTrainer withCountOfComponents(int numberOfComponents) {
+        A.ensure(numberOfComponents > 0, "Number of components in GMM cannot equal 0");
```
[the remark this answers is not in the document]
Comment (avplatonov) [reply modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/GmmTrainer.java]: :+1:

File: modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/GmmTrainer.java [hunk 13]
```diff
@@ -0,0 +1,346 @@
...
+    /**
+     * Sets initial means.
+     *
+     * @param means Initial means for clusters.
+     * @return trainer.
+     */
+    public GmmTrainer withInitialMeans(List<Vector> means) {
+        A.notEmpty(means, "GMM should starts with non empty initial components list");
+
+        this.initialMeans = means.stream().toArray(Vector[]::new);
```
[the remark this answers is not in the document]
Comment (avplatonov) [reply modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/GmmTrainer.java]: thnx

File: modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/GmmTrainer.java [hunk 14]
```diff
@@ -0,0 +1,346 @@
...
+                    randomMeansSets.stream().mapToInt(List::size).sum() >= countOfComponents,
+                    "There is not enough data in dataset for select N random means"
+                );
+
+                initialMeans = new Vector[countOfComponents];
+                int j = 0;
+                for (int i = 0; i < countOfComponents; ) {
+                    List<Vector> randomMeansPart = randomMeansSets.get(j);
+                    if (!randomMeansPart.isEmpty()) {
+                        initialMeans[i] = randomMeansPart.remove(0);
```
[the remark this answers is not in the document]
Comment (avplatonov) [reply modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/GmmTrainer.java]: Yes, I thought about it. But I was lazy)

File: modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/GmmTrainer.java [hunk 15]
```diff
@@ -0,0 +1,346 @@
...
+
+                    j = (j + 1) % randomMeansSets.size();
+                }
+            }
+
+            dataset.compute(data -> GmmPartitionData.estimateLikelihoodClusters(data, initialMeans));
+
+            List<Matrix> initialCovs = CovarianceMatricesAggregator.computeCovariances(
+                dataset,
+                VectorUtils.of(DoubleStream.generate(() -> 1. / countOfComponents).limit(countOfComponents).toArray()),
```
[the remark this answers is not in the document]
Comment (avplatonov) [reply modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/GmmTrainer.java]: thank you

File: modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/GmmTrainer.java [hunk 16]
```diff
@@ -0,0 +1,346 @@
...
+            res.add(new MultivariateGaussianDistribution(means[i], covs.get(i)));
+
+        return res;
+    }
+
+    /**
+     * Check algorithm covergency. If it's true then algorithm stops.
+     *
+     * @param oldModel Old model.
+     * @param newModel New model.
```
[the remark this answers is not in the document]
Comment (avplatonov) [reply modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/GmmTrainer.java]: :+1:

File: modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/GmmTrainer.java [hunk 17]
```diff
@@ -0,0 +1,346 @@
...
+    private boolean isConverged(GmmModel oldModel, GmmModel newModel) {
+        A.ensure(oldModel.countOfComponents() == newModel.countOfComponents(),
+            "oldModel.countOfComponents() == newModel.countOfComponents()");
+
+        boolean isConverged = true;
+        for (int i = 0; i < oldModel.countOfComponents(); i++) {
+            MultivariateGaussianDistribution d1 = oldModel.distributions().get(i);
+            MultivariateGaussianDistribution d2 = newModel.distributions().get(i);
+
+            isConverged = isConverged && Math.sqrt(d1.mean().getDistanceSquared(d2.mean())) < eps;
```
[the remark this answers is not in the document]
Comment (avplatonov) [reply modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/GmmTrainer.java]: :1

File: modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/MeanWithClusterProbAggregator.java [hunk 18]
```diff
@@ -0,0 +1,205 @@
...
+    private static final long serialVersionUID = 2700985110021774629L;
+
+    /** Weighted sum of vectors. */
+    private Vector weightedXsSum;
+
+    /** P(C|xi) sum. */
+    private double pcxiSum;
+
+    /** Aggregated partition data size. */
+    private int N;
```
[the remark this answers is not in the document]
Comment (avplatonov) [reply modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/MeanWithClusterProbAggregator.java]: fixed

File: modules/ml/src/main/java/org/apache/ignite/ml/math/primitives/matrix/AbstractMatrix.java (hunk 1 above)
Comment (avplatonov) [reply modules/ml/src/main/java/org/apache/ignite/ml/math/primitives/matrix/AbstractMatrix.java]: Yeah, great idea. Replace destroyable iface to autocloseable.

File: modules/ml/src/main/java/org/apache/ignite/ml/math/stat/DistributionMixture.java (hunk 2 above)
Comment (avplatonov) [reply modules/ml/src/main/java/org/apache/ignite/ml/math/stat/DistributionMixture.java]: :+1:

File: modules/ml/src/main/java/org/apache/ignite/ml/nn/ReplicatedVectorMatrix.java (hunk 3 above)
Comment (avplatonov) [reply modules/ml/src/main/java/org/apache/ignite/ml/nn/ReplicatedVectorMatrix.java]: fixed
